[WIP] Run tests in docker

a-magdy opened the pull request:
This PR aims at running monstache tests in docker, paving the way for CI and quicker and easier contribution process

These files are in root directory to speed up testing, but once the changes are approved as a concept, we'll work on cleaning up the code, moving it to other folders, and documenting everything

To run use the script: `run-tests.sh`

Currently it is failing in:

```bash
c-monstache-test-runner | --- FAIL: TestDropCollection (3.06s)
c-monstache-test-runner |       monstache_test.go:298: elastic: Error 404 (Not Found): no such index [type=index_not_found_exception]
c-monstache-test-runner | FAIL
c-monstache-test-runner | exit status 1
c-monstache-test-runner | FAIL  app     15.219s
```

I am not sure what is the problem, as it succeeds in connecting to both mongodb and elasticsearch

rwynn commented:
I normally run the monstache process for the test with no explicit config. Here you would need to at least config the 2 url. 
I don’t usually use regex or direct reads. Presence of direct reads currently causes the flush timeout to increase to 5 seconds. You might want to explicitly set elasticsearch-max-seconds to 1 to prevent that.

rwynn commented:
And settings elasticsearch-max-docs to 1 would also help since the test only pushes a couple changes.

a-magdy commented:
I noticed that I had an extra `.` in the namespaces regex, fixed it and tests started passing

Then I started seeing another error:

```
=== RUN   TestDropDatabase
--- FAIL: TestDropDatabase (40.05s)
	monstache_test.go:269: clientsearch index not deleted
=== RUN   TestDropCollection
--- FAIL: TestDropCollection (40.04s)
	monstache_test.go:309: clientsearch index not deleted
FAIL
exit status 1
FAIL	app	180.233s
```

So I changed both settings (elasticsearch-max-[seconds&docs] to 1, but I am still getting the last error

can you please try it locally on your machine

rwynn commented:
Well, for one that error should say **elasticsearch** index not deleted.  I think I must have done a search/replace at some point and changed all elastic to client, even in the error messages!

So, I think the problem is that the regex is not accounting for the namespace `test.$cmd` which is what would be produced in the oplog when a table is dropped.  

I have a note on that in the docs but it's easily missed.
https://rwynn.github.io/monstache-site/advanced/#namespaces

So I think the drop database/collection events are being ignored because of the regex filter.

a-magdy commented:
Hmm, so what should I do, I commented the setting: `namespace-regex` completely, didn't work
change it to `'^*(test.)*$'`, `'^*test*$'` & `'^*$'` & `'*'` and all of them failed for different reasons

I am blocked in this state for few hours now, it would be nice if you could run locally and fix the monstache config the way you know it works

And at this note, I think splitting regex patterns for dropping would be easier to grasp if the command is different in mongo oplog (atleast allow the option to override the drop regex for advanced usages)

rwynn commented:
I think I found out why the tests are failing.  There was a change to `gtm` recently which is causing db and collection `drops` to be applied out of sequence. Each test drops the `test.test` db at the start, and this operation is getting applied to Elasticsearch in the **middle** of the test instead of the start and that's why we're seeing index not found errors.  I will put some commits later today (to the code not the tests) to fix this. 

This compose setup is great by the way.  Thanks!

rwynn commented:
I should say it passes when the `namespace-regex` is removed completely.

a-magdy commented:
@rwynn That is great, thank you

I'll continue working on it, sorry for holding it for a while
I'll have to change the way it runs to make the tests exists with 0 after a successful run

You are right, I removed `namespace-regex` and it work 🎉

rwynn commented:
@a-magdy sorry I just noticed you last comment.  I had been working on this and pushed 2 commits for it.

https://github.com/rwynn/monstache/commit/28a1eca4ad1f80fd7377347357da70f5d7f9fbb6

https://github.com/rwynn/monstache/commit/bd92c74c7b88ff4340d00a723292f57c2d0ba1ce

Sorry for any inconvenience.

a-magdy commented:
Np :) it is nice that you updated it yourself, I'll compare the changes, and will try to find a way, so that the tests exit with 0 (success) or 1 (failure) after it runs, instead of having to look at the logs

Also it'll be required to exit in the CI process when it comes

a-magdy commented:
And sorry for delaying the work, I am completely hammered with work these days (including weekends) 🤕

rwynn commented:
No problem.  You've done a ton of work already that has pushed me to get this automation set up.  I've just pushed integration with Travis CI, so the docker compose stuff that you did should be tested and linked from the github home page now for both the master and rel3 branches.  The rel3 branch uses a 5.x docker image from Elastic since that is what that branch is supposed to be going against.

a-magdy commented:
Great, it is nice seeing CI working :) 

I removed the files in the root directory and made some small changes instead, if you approve, I think we can close this PR
